universe/Building.h

```diff
@@ -214,9 +214,8 @@ class BuildingTypeManager {
 
 private:
     BuildingTypeManager();
-    ~BuildingTypeManager();
 
-    std::map<std::string, BuildingType*> m_building_types;
+    std::map<std::string, std::unique_ptr<BuildingType>> m_building_types;
```

Comment:
Returning `unique_ptr` from the parser is correct.  It is explicit in the function signature that the parser does not retain any stake in the parsed object.  Returning `shared_ptr` implies otherwise.

This PR is just focused on fixing the parser.

The `BuildingTypeManager` never passes ownership of the building types to any other code.  It is statically allocated and never destroyed so its lifetime exceeds all of its calling functions.  It is correct for the building manager to return raw pointers.

Converting a received `unique_ptr` from a generator into a `shared_ptr` is as simple as constructing from the `unique_ptr`:
````
unique_ptr<T> generatorT() {
    return make_unique<T>();
}

main() {
    shared_ptr<T> receive_T = generatorT();
}
````  
It is one line of code, not more lines of code than not doing the conversion. It increases code clarity because the transfer of ownership is explicit. 

Here is further exposition to clarify what I think are common conventions for smart pointers in C++:
- `unique_ptr<T>` - I am the sole owner of this `T`.  If I move this `unique_ptr<T>` to you then you become the sole owner of `T`.  There is no overhead (relative to a raw pointer) to dereferencing, moving, copying or destroying this `unique_ptr<T>`.  There is possibly a memory allocation overhead to storing a custom destructor for this `T`
- `shared_ptr<T>` - I share ownership of this `T`. If I move/copy this `shared_ptr<T>` then you also share ownership. There is no overhead to dereferencing or moving this `shared_<T>`.  Copying or destroying this `shared_ptr<T>` requires an extra memory dereference to the control block and an atomic increment/decrement.  Constructing this `shared_ptr<T>` requires allocating additional memory for the control block. If constructed with `make_shared` then the control block may be contiguous with the `T`.
- `weak_ptr<T>`  - I can observe a `shared_ptr<T>`, but I do not own it nor can I prevent it from being destroyed.  "Dereferencing", constructing, copying or destroying `weak_ptr<T>` all have overhead. 
- raw `T*`  - Some other entity is managing the memory for this object.  I know, from program semantics, that the lifetime of this objects exceeds my use.  There is no overhead to any operations.